Fully Revise docs with latest API for 1.2.0 (#22)
wisp-forest/docs: docs/accessories/developer/api_fundamentals.md

### Adding a Custom Accessory Renderer

With creating a custom `Accessory` comes with the ability to add some rendering that is attached to the player. You can do such by creating a custom instance of `AccessoryRenderer` and registering such using the `AccessoriesRendererRegistry.registerRenderer` method. More information about the API available to developers can be found in the rendering API section. {TODO: WIP ADD LINK TO SUCH}

!!! warning "Default Rendering Behavior"
    By default as discussed within the Rendering API section of the docs, All Accessories will use `DefaultAccessoryRenderer` unless a custom `AccessoryRenderer` is registered or no renderer is registered using `AccessoriesRendererRegistry#registerNoRenderer` as such is meant to have some rendering.

### Creating a custom Slot Predicate

In cases where a Tag may not the be the best solution or using Data Components on your ItemStack may not be suitable, one can create a custom `SlotBasedPredicate` where you can have more general control of allowing `Accessory` entries from being equipped to a given slot. 

> This is useful for dynamically accessory types that may have multiple iterations that would make it difficult to use Tag without possible issues.

Reviewer: the use of "dynamically" is wrong here but i'm not sure if its meant to just say "dynamic" or if there's meant to be another word between "dynamically" and "accessory"

!!! info "Entity Based Predicate"
    With the release of 1.2.0, `EntityBasedPredicate` exists as a way to have entity context when evaluating if a given Accessory should be equipped within a given slot. It is recommend that this is used only if needed as it may be difficult to evaluate for things like crafting where the target entity may differ.

### Accessing The Capability

Within Accessories there is the `AccessoriesCapability` which is used to access the slot containers, adjust slot attributes, or query the inventory for various possible item stacks. 

Any interaction involving the entity would be found here if they have a `AccessoriesCapability` as the entity needs to have slots bound to such to be able to get the capability. This means that your logic should revolve around the idea that not every entity can be capable of being interacted with unless they have slots bound to them.

### Dealing with the Container

> For each slot there will be a `AccessoriesContainer` instance if the given Entity has the slot bound to such. That means that it has the possibility to not exist for every entity meaning you should confirm such to exist in all cases.

Reviewer: "a" -> "an"

---
title: API Fundamentals
project: accessories
---

### Creating an Accessory

You can either create an Accessory by implementing the `Accessory` interface or by extending the `AccessoryItem` which implements the `Accessory` interface. 

> The first requires that you register your custom `Accessory` to the given `Item` instance using the `AccessoryRegistry#register` method while the later `AccessoryItem` will register the accessory when the class is instantiated

Author: Me when I meant "the latter" and not later but frankly is better phrasing on how `AccessoryItem` works

> This `Accessory` interface contains various functions that can be preformed like `tick`ing, controlling equipablity, listening to when equipment is changed, adding attribute modifiers dynamically or statically, adding custom tooltip info, and more.

Reviewer 2: ```suggestion
This `Accessory` interface contains various functions that can be performed. Some examples include `tick`ing, controlling equipability, listening to when equipment is changed, adding attribute modifiers (dynamically or statically), and adding custom tooltip info.
```

### Adding a Custom Accessory Renderer

> Creating a custom `Accessory` comes with the ability to add some rendering that is attached to the player. You can do such by creating a custom instance of `AccessoryRenderer` and registering such using the `AccessoriesRendererRegistry.registerRenderer` method. More information about the API available to developers can be found in the [rendering API section](./rendering_api.md).

Reviewer 2: ```suggestion
Creating a custom `Accessory` comes with the ability to add some rendering that is attached to the player. You can do so by creating a custom instance of `AccessoryRenderer` and registering it using the `AccessoriesRendererRegistry.registerRenderer` method. More information about the rendering API be found in the [rendering API section](./rendering_api.md).
```

Author: Yea good idea to get rid of such as it seems such is used quite often here for some such reason idk why such would such be used in instead of such other words that you used instead of such...

!!! warning "Default Rendering Behavior"

>     By default as discussed within the Rendering API section of the docs, all Accessories will use `DefaultAccessoryRenderer` unless a custom `AccessoryRenderer` is registered or no renderer is registered using `AccessoriesRendererRegistry#registerNoRenderer`.

Reviewer 2: ```suggestion
    By default within the Rendering API: all Accessories will use `DefaultAccessoryRenderer`, unless a custom `AccessoryRenderer` is registered, or an explicit "no renderer" is registered using `AccessoriesRendererRegistry#registerNoRenderer`.
```

### Creating a custom Slot Predicate

> In cases where a Tag may not be the best solution or using Data Components on your ItemStack may not be suitable, one can create a custom `SlotBasedPredicate` where you can have more general control of allowing `Accessory` entries to be equipped to a given slot. 

Reviewer 2: ```suggestion
In cases where neither a Tag or using Data Components on your Item Stack are suitable, you may create a custom `SlotBasedPredicate` where you can have more control over what `Accessory` entries are allowed to be equipped in a given slot. 
```

This is useful for dynamic accessory types that may have multiple iterations that would make it difficult to use a Item Tag without possible issues.

!!! info "Entity Based Predicate"

>     With the release of 1.2.0, `EntityBasedPredicate` exists as a way to have entity context when evaluating if a given Accessory should be equipped within a given slot. It is recommend that this is used only if needed as it may be difficult to evaluate for things like crafting where the target entity may differ.

Reviewer 2: ```suggestion
    With the release of 1.2.0, `EntityBasedPredicate` exists as a way to have entity context when evaluating if a given Accessory should be equipped within a given slot. It is recommend that this is used only if needed, as it can be difficult to evaluate for cases like crafting where the target entity may differ.
```

### Accessing The Capability

Within Accessories there is the `AccessoriesCapability` which is used to access the slot containers, adjust slot attributes, or query the inventory for various possible item stacks. 

> Any interaction involving the entity would be found here if they have a `AccessoriesCapability` as the entity needs to have slots bound to such to be able to get the capability. This means that your logic should revolve around the idea that not every entity can be capable of being interacted with unless they have slots bound to them.

Reviewer 2: ```suggestion
Any interaction involving an entity can only be done if they have an `AccessoriesCapability`, which implies that the entity has slots bound to it. This means that your logic should revolve around the idea that not every entity can be capable of being interacted with unless they have slots bound to them.
```
> This means that your logic should revolve around the idea that not every entity can be capable of being interacted with unless they have slots bound to them.

What does this mean exactly? Does this essentially boil down to "check for the capability"?

Author: Yes as I guess that may be the general statement but I want to explain it in a way of why it may not exist

### Adjusting Accessory Equipping

> A considerable difference between other Accessory API's is that `canEquip` is checked on loading Accessory data and revalidate on reload requiring that you use `isAnotherEquipped` found on the given entities `AccessoriesCapability` to prevent duplicate Accessory items from being equipped. Such requires the current `ItemStack` and `SlotReference` of the current location to be ignored when checking for another equipped accessory.

Reviewer 2: ```suggestion
A considerable difference between other Accessory API's is that `canEquip` is checked when Accessory data is loaded, and revalidated on reload. This requires you to use `isAnotherEquipped` found on the given entities `AccessoriesCapability` to prevent duplicate Accessory items from being equipped. This requires the current `ItemStack` and `SlotReference` of the current location to be ignored when checking for another equipped accessory.
```
> "... current **location**"

To me this seems vague - what does location imply here? The Slot Reference? The Accessory? Needs to be a bit more specific.

Author: Where the stack is located that being the current `SlotReference` instance given i.e, the entity's slot container index stack location

Author: Yea, that seems good.

### Dealing with the Container

> For each slot there will be an `AccessoriesContainer` instance if the given Entity has the slot bound to such. That means that it has the possibility to not exist for every entity meaning you should confirm such to exist in all cases.

Reviewer 2: ```suggestion
For each slot on an entity - there will be an `AccessoriesContainer` instance if the given Entity has the slot bound to it. This means that the specified slot may not exist on the entity, meaning you should check for this.
```
This suggestion could be further improved, as it is similarily vague. It might be obvious if you check the API, but does this mean you have to check for `null` or some `EMPTY` value?

Author: It means that it may be possible be null for the given entity either for the given type or for all types based on adjustments within data packs files

> This object allows for getting information like the current equipped accessories, render toggles for given accessories and managing the slot attributes stored. It mainly not recommend to directly use the container to look for Accessories but instead use the methods on `AccessoriesCapability` instead.

Reviewer 2: ```suggestion
This object allows for getting information like the current equipped accessories, render toggles for given accessories and managing the slot attributes stored. It is not recommended to directly use the container to look for Accessories. Use the methods on the `AccessoriesCapability` instead.
```
Is there a reason these methods which should not be used are accessible in the API?

Author: It's mainly down to the fact that it's undecided whether access to such is a good idea directly from the `AccessoryContainer`, but it may be useful for people implementing a transfer API to access the direct Inventory/Container instances for the Accessory Container.
